capycli/bom/create_components.py

```diff
@@ -299,6 +299,8 @@ def update_release(self, cx_comp: Component, release_data: Dict[str, Any]) -> No
                 print_yellow(
                     "    WARNING: SW360 source URL", release_data["sourceCodeDownloadurl"],
                     "differs from BOM URL", data["sourceCodeDownloadurl"])
+                if data["sourceCodeDownloadurl"].endswith(('zip', 'tgz', 'tar.gz', 'tar')):
+                    update_data["sourceCodeDownloadurl"] = data["sourceCodeDownloadurl"]
```

Comment:
Sorry for being late to this review, but can somebody please explain me this change? I understand it as "if URL in BOM looks somehow correct, always overwrite URL in SW360". Do we really want this? Or do I get it wrong here?

Reply:
As elaborated in the other comments, the source code detection often doesn't find correct URLs, or the package's metadata isn't correct (my experience so far was that JS projects perform the worst).
As a consequence, I've also observed that there are often existing releases in SW360 that have existing source code attachments with wrong file types (e.g., .git), which wouldn't be corrected by the capycli due to the attachment already existing, even though the capycli would now (with the introduced changes) often find a valid source code download URL.

Yes, the check is broader than overriding just the existing URLs ending with .git (for which this was introduced), but in my testing, I haven't observed negative side effects.

I would gladly hear your opinion. If you think this could introduce negative side effects, we can make it more specific to target clearly wrong source code URLs (e.g., ending with .git).

Reply:
Ok, I tested it now and it will now *always* update existing source URLs given that the URL in the BOM ends with .tar.gz oder .zip. But probably the existing entry in SW360 is more correct than my BOM? Apart from that, the user would probably not expect this.

In Debian, for example, I found that many solutions generate somehow wrong source URLs, e.g. pointing to orig.tar.gz or orig.tar.xz packages. So in some cases, existing, valid URLs would be replaced by invalid ones, in others not (because .tar.gz is on the whitelist, .tar.xz is not).

Apart from that, in Debian, we use permanent URLs like https://snapshot.debian.org/archive/debian/<timestamp>/..., so nearly every user will have a different URL because same package can be found in different snapshots. So this could trigger *lots* of unnecessary changes and thus moderation requests when handling large BOMs. (Actually, it would not for my concrete example, as correct URLs end with .tsc, only incorrect ones would cause this...).

I think the code should at least be narrowed to update only URLs you consider invalid, but I'm really not sure if hard-coding a fix for a specific JS issue in the general code paths is a good idea. I personally never stumbled over such wrong entries in other SW360 entries. So my suggestion would be to roll-back these changes, open an issue about your specific problem and discuss possible solutions there.

Reply:
Note that I just create #96 so we have a testcase for these changes and we can shape it according to a decision how to handle your issue. We might think about adding test-case for the other changes introduced here, too.